pkg/api/api.go

```diff
@@ -293,112 +263,22 @@ func (a *API) RegisterCompactor(c *compactor.Compactor) {
-		func() config.Config { return config.Config{} },
-		map[string]string{}, // TODO: include configuration flags
-		v1.GlobalURLOptions{},
-		func(f http.HandlerFunc) http.HandlerFunc { return f },
-		nil,   // Only needed for admin APIs.
-		"",    // This is for snapshots, which is disabled when admin APIs are disabled. Hence empty.
-		false, // Disable admin APIs.
-		a.logger,
-		func(context.Context) v1.RulesRetriever { return &querier.DummyRulesRetriever{} },
-		0, 0, 0, // Remote read samples and concurrency limit.
-		regexp.MustCompile(".*"),
-		func() (v1.RuntimeInfo, error) { return v1.RuntimeInfo{}, errors.New("not implemented") },
-		&v1.PrometheusVersion{},
-	)
-
+) {
 	// these routes are always registered to the default server
 	a.RegisterRoute("/api/v1/user_stats", http.HandlerFunc(distributor.UserStatsHandler), true, "GET")
 	a.RegisterRoute("/api/v1/chunks", querier.ChunksHandler(queryable), true, "GET")
 
 	a.RegisterRoute(a.cfg.LegacyHTTPPrefix+"/user_stats", http.HandlerFunc(distributor.UserStatsHandler), true, "GET")
 	a.RegisterRoute(a.cfg.LegacyHTTPPrefix+"/chunks", querier.ChunksHandler(queryable), true, "GET")
-
-	// these routes are either registered the default server OR to an internal mux.  The internal mux is
-	// for use in a single binary mode when both the query frontend and the querier would attempt to claim these routes
-	// TODO:  Add support to expose querier paths with a configurable prefix in single binary mode.
-	router := mux.NewRouter()
-	if registerRoutesExternally {
-		router = a.server.HTTP
-	}
-
-	// Use a separate metric for the querier in order to differentiate requests from the query-frontend when
-	// running Cortex as a single binary.
-	inst := middleware.Instrument{
-		RouteMatcher:     router,
-		Duration:         querierRequestDuration,
-		RequestBodySize:  receivedMessageSize,
-		ResponseBodySize: sentMessageSize,
-		InflightRequests: inflightRequests,
-	}
-
-	promRouter := route.New().WithPrefix(a.cfg.ServerPrefix + a.cfg.PrometheusHTTPPrefix + "/api/v1")
-	api.Register(promRouter)
-	cacheGenHeaderMiddleware := getHTTPCacheGenNumberHeaderSetterMiddleware(tombstonesLoader)
-	promHandler := fakeRemoteAddr(inst.Wrap(cacheGenHeaderMiddleware.Wrap(promRouter)))
-
-	a.registerRouteWithRouter(router, a.cfg.PrometheusHTTPPrefix+"/api/v1/read", querier.RemoteReadHandler(queryable), true, "POST")
-	a.registerRouteWithRouter(router, a.cfg.PrometheusHTTPPrefix+"/api/v1/query", promHandler, true, "GET", "POST")
-	a.registerRouteWithRouter(router, a.cfg.PrometheusHTTPPrefix+"/api/v1/query_range", promHandler, true, "GET", "POST")
-	a.registerRouteWithRouter(router, a.cfg.PrometheusHTTPPrefix+"/api/v1/labels", promHandler, true, "GET", "POST")
-	a.registerRouteWithRouter(router, a.cfg.PrometheusHTTPPrefix+"/api/v1/label/{name}/values", promHandler, true, "GET")
-	a.registerRouteWithRouter(router, a.cfg.PrometheusHTTPPrefix+"/api/v1/series", promHandler, true, "GET", "POST", "DELETE")
```

Comment:
Who register these now?

Reply:
These routes are registered with the two routers when we call [api.Register](https://github.com/prometheus/prometheus/blob/master/web/api/v1/api.go#L259) on them. We end up having to register two routers because the upstream prometheus [route library](https://github.com/prometheus/common/blob/master/route/route.go#L88) and registration function requires us to register separately for each path prefix.

Previously we explicitly registered every route in the Prometheus API to query handler. However, now we only explicitly register the `metadata` and `read` routes because they aren't handled by the upstream API struct. For the rest we register the prefix and since the `promHandler` and `legacyPromHandler` are themselves routers, the requests should be forwarded correctly. 

Also I should note that the only way to hit these routes is by a request  being routed to the `internalQuerierHandler` which can only happen if it hit's the routes registered in `RegisterQueryAPI`, so no new routes should have been exposed and no existing routes should have been removed. This PR just simplified the router that already was only acting as a middleman to the upstream Prometheus routers we already registered with the API struct.